File: mc_pygame_controller/chain_utils.py
from dataclasses import dataclass, field, replace
import json
from typing import List, Dict, Union, Any, Optional, Tuple, Type
from functools import wraps
import inspect
import os
import base64
import httpx
import mimetypes
from openai import AsyncOpenAI
from pydantic import BaseModel
import asyncio
from .mcp_server import Server, create_tool_functions, Configuration



def chain_method(func):
    """Decorator for chain methods to ensure they return self for chaining."""
    @wraps(func)
    def wrapper(self, *args, **kwargs):
        result = func(self, *args, **kwargs)
        if result is None:
            return self
        return result
    return wrapper



async def encode_base64_content_from_url(content_url: str) -> str:
    """Asynchronously fetch content from a URL and encode it in base64."""

    async with httpx.AsyncClient() as client:
        response = await client.get(content_url)
        response.raise_for_status()
        result = base64.b64encode(response.content).decode("utf-8")

    return result


async def _resolve_multimodal_args(args: Dict[str, Any]) -> Dict[str, Any]:
    """Convert any URL fields in tool arguments to base64-encoded strings."""

    resolved = {}
    for key, value in args.items():
        if isinstance(value, list):
            resolved[key] = [
                (
                    await encode_base64_content_from_url(v)
                    if isinstance(v, str) and v.startswith("http")
                    else v
                )
                for v in value
            ]
        elif isinstance(value, str) and value.startswith("http"):
            resolved[key] = await encode_base64_content_from_url(value)
        else:
            resolved[key] = value
    return resolved


async def _encode_to_data_uri(source: str, mime_type: Optional[str] = None) -> str:
    """Encode a local file or remote URL to a base64 data URI."""

    if source.startswith("http"):
        async with httpx.AsyncClient() as client:
            response = await client.get(source)
            response.raise_for_status()
            content = response.content
            if not mime_type:
                mime_type = response.headers.get("content-type")
    else:
        with open(source, "rb") as f:
            content = f.read()
        if not mime_type:
            mime_type = mimetypes.guess_type(source)[0]

    mime_type = mime_type or "application/octet-stream"
    encoded = base64.b64encode(content).decode("utf-8")
    return f"data:{mime_type};base64,{encoded}"


async def _resolve_multimodal_output(output: Any) -> Any:
    """Convert MCP multimodal content to OpenAI message format."""

    if isinstance(output, dict) and "content" in output:
        # Handle MCP multimodal content structure
        content_items = output["content"]

        # If there's only text content, return just the text
        text_items = [item for item in content_items if item.get("type") == "text"]
        image_items = [item for item in content_items if item.get("type") == "image"]

        if len(content_items) == 1 and content_items[0].get("type") == "text":
            return content_items[0].get("text", "")

        # For multimodal content, we'll return structured data that the chain can handle
        result = []

        for item in content_items:
            if item.get("type") == "text" and item.get("text"):
                result.append({"type": "text", "text": item["text"]})
            elif item.get("type") == "image" and item.get("data"):
                # Convert to OpenAI format
                mime_type = item.get("mimeType", "image/png")
                data_uri = f"data:{mime_type};base64,{item['data']}"
                result.append({"type": "image_url", "image_url": {"url": data_uri}})

        return {"multimodal_content": result}

    elif isinstance(output, str):
        if output.startswith("http") or os.path.exists(output):
            return await _encode_to_data_uri(output)
        return output
    elif isinstance(output, list):
        return [await _resolve_multimodal_output(v) for v in output]
    elif isinstance(output, dict):
        return {k: await _resolve_multimodal_output(v) for k, v in output.items()}
    else:
        return output


def chain_method(func):
    """Decorator to convert a function into a chainable method that supports
    both synchronous and asynchronous functions."""

    if inspect.iscoroutinefunction(func):

        @wraps(func)
        async def async_wrapper(self, *args, **kwargs):
            return await func(self, *args, **kwargs)

        return async_wrapper
    else:

        @wraps(func)
        def wrapper(self, *args, **kwargs):
            return func(self, *args, **kwargs)

        return wrapper

File: mc_typescript_controller/src/chain_utils.ts
/**
 * @file chain_utils.ts
 * Utilities for handling multimodal content, data encoding, and potentially
 * for more complex chain operations in the future.
 */

import axios from 'axios';
import *_fs from 'fs'; // Renamed to avoid conflict with fs namespace if any
import _path from 'path'; // Renamed
import *mime from 'mime-types'; // For guessing MIME types

const fs = _fs.promises; // Use promises API for async file operations

/**
 * Asynchronously fetches content from a URL and encodes it in base64.
 * @param content_url - The URL to fetch content from.
 * @returns A Promise that resolves to the base64 encoded string.
 */
export async function encode_base64_content_from_url(content_url: string): Promise<string> {
    try {
        const response = await axios.get(content_url, { responseType: 'arraybuffer' });
        if (response.status !== 200) {
            throw new Error(`Failed to fetch URL: ${content_url}, status: ${response.status}`);
        }
        return Buffer.from(response.data).toString('base64');
    } catch (error: any) {
        console.error(`Error fetching or encoding URL ${content_url}: ${error.message}`);
        throw error;
    }
}

/**
 * Recursively processes arguments: if a string value starts with "http",
 * it's treated as a URL and its content is fetched and base64 encoded.
 * Lists of strings are also processed.
 * @param args - The arguments object to resolve.
 * @returns A Promise that resolves to the arguments object with URLs replaced by base64 content.
 */
export async function _resolve_multimodal_args(args: Record<string, any>): Promise<Record<string, any>> {
    const resolved: Record<string, any> = {};
    for (const key in args) {
        if (Object.prototype.hasOwnProperty.call(args, key)) {
            const value = args[key];
            if (Array.isArray(value)) {
                resolved[key] = await Promise.all(
                    value.map(async (v) => {
                        if (typeof v === 'string' && (v.startsWith("http://") || v.startsWith("https://"))) {
                            return encode_base64_content_from_url(v);
                        }
                        return v;
                    })
                );
            } else if (typeof value === 'string' && (value.startsWith("http://") || value.startsWith("https://"))) {
                resolved[key] = await encode_base64_content_from_url(value);
            } else {
                resolved[key] = value;
            }
        }
    }
    return resolved;
}

/**
 * Encodes a local file or remote URL to a base64 data URI.
 * @param source - The path to a local file or an HTTP(S) URL.
 * @param explicit_mime_type - Optional explicit MIME type to use. If not provided, it's guessed.
 * @returns A Promise that resolves to the base64 data URI.
 */
export async function _encode_to_data_uri(source: string, explicit_mime_type?: string): Promise<string> {
    let content: Buffer;
    let mime_type = explicit_mime_type;

    if (source.startsWith("http://") || source.startsWith("https://")) {
        try {
            const response = await axios.get(source, { responseType: 'arraybuffer' });
            if (response.status !== 200) {
                throw new Error(`Failed to fetch URL: ${source}, status: ${response.status}`);
            }
            content = Buffer.from(response.data);
            if (!mime_type) {
                mime_type = response.headers['content-type']?.split(';')[0]; // Get content type, remove parameters like charset
            }
        } catch (error: any) {
            console.error(`Error fetching URL ${source}: ${error.message}`);
            throw error;
        }
    } else {
        try {
            if (!_fs.existsSync(source)) { // Use synchronous existsSync for quick check before async read
                 throw new Error(`File not found: ${source}`);
            }
            content = await fs.readFile(source);
            if (!mime_type) {
                mime_type = _mime.lookup(source) || undefined;
            }
        } catch (error: any) {
            console.error(`Error reading file ${source}: ${error.message}`);
            throw error;
        }
    }

    mime_type = mime_type || "application/octet-stream"; // Default MIME type
    const encoded = content.toString('base64');
    return `data:${mime_type};base64,${encoded}`;
}


/**
 * Resolves multimodal output content. If the output is an MCP-style multimodal content
 * structure, it converts image data to data URIs for OpenAI compatibility.
 * If output is a string URL or local path, it's encoded to a data URI.
 * Handles nested structures (lists, dicts/objects).
 * @param output - The output data to resolve.
 * @returns A Promise that resolves to the processed output.
 */
export async function _resolve_multimodal_output(output: any): Promise<any> {
    if (typeof output === 'string') {
        if (output.startsWith("http://") || output.startsWith("https://") || _fs.existsSync(output)) {
             try {
                return await _encode_to_data_uri(output);
             } catch(e: any) {
                console.warn(`Could not encode string source "${output}" to data URI: ${e.message}. Returning as is.`);
                return output; // Return original string if encoding fails
             }
        }
        return output; // Return other strings as is
    } else if (Array.isArray(output)) {
        return Promise.all(output.map(item => _resolve_multimodal_output(item)));
    } else if (output && typeof output === 'object' && output !== null) {
        if (output.content && Array.isArray(output.content)) { // MCP multimodal structure
            const new_content_items: any[] = [];
            for (const item of output.content) {
                if (item && item.type === "text" && typeof item.text === 'string') {
                    new_content_items.push({ type: "text", text: item.text });
                } else if (item && item.type === "image" && typeof item.data === 'string') {
                    const mimeType = item.mimeType || "image/png"; // Default to PNG if not specified
                    const dataUri = `data:${mimeType};base64,${item.data}`;
                    new_content_items.push({ type: "image_url", image_url: { url: dataUri } });
                } else {
                    new_content_items.push(item); // Pass through unknown items
                }
            }
            // If only text, simplify to string, otherwise return OpenAI structure
            if (new_content_items.length === 1 && new_content_items[0].type === "text") {
                return new_content_items[0].text;
            }
            return { multimodal_content: new_content_items }; // Retain this structure for OpenAI tools
        } else { // Generic object, recurse
            const resolved_obj: Record<string, any> = {};
            for (const key in output) {
                if (Object.prototype.hasOwnProperty.call(output, key)) {
                    resolved_obj[key] = await _resolve_multimodal_output(output[key]);
                }
            }
            return resolved_obj;
        }
    }
    return output; // Return primitives and null/undefined as is
}


/**
 * Higher-order function to wrap a method, ensuring it can be used in a fluent chaining style.
 * This is a simplified interpretation of the Python `chain_method` decorator.
 * It doesn't fully replicate Python decorator behavior but provides basic async/sync wrapping.
 *
 * @param func - The function to wrap.
 * @returns The wrapped function.
 */
export function chainMethod<T extends (...args: any[]) => any>(func: T): T {
    // This HOF doesn't modify 'this' or automatically return 'this'.
    // The original Python decorator's implicit 'return self' is not replicated here.
    // The calling code would need to ensure it returns 'this' if that's the desired chaining behavior.

    // Check if it's an async function
    const isAsync = func.constructor.name === "AsyncFunction";

    if (isAsync) {
        return (async function(this: any, ...args: Parameters<T>): Promise<ReturnType<T>> {
            // 'this' context is preserved by arrow function or by how it's called
            return await func.apply(this, args);
        } as T);
    } else {
        return (function(this: any, ...args: Parameters<T>): ReturnType<T> {
            return func.apply(this, args);
        } as T);
    }
}

// Example of how it might be used (conceptual):
// class MyChain {
//     @chainMethod // If using TS decorators and the decorator is defined for that
//     myMethod(value: string) {
//         console.log(value);
//         return this; // Explicitly return this for chaining
//     }
//
//     // Or without decorators:
//     myOtherMethod = chainMethod(function(this: MyChain, value: number) {
//         console.log(value);
//         return this;
//     });
// }
